### PR TITLE
mesa: update to 24.1.0

### PR DESCRIPTION
**Pull request by heitbaum**
announcement:
- https://lists.freedesktop.org/archives/mesa-announce/2024-May/000762.html

release notes:
- https://gitlab.freedesktop.org/mesa/mesa/-/blob/24.1/docs/relnotes/24.1.0.rst

**iris** driver included in mesa-24.1.0 requires intel_clc for build (thus required updates to llvm, spirv-, libdrm)
Some ref regarding libdrm:
- https://gitlab.freedesktop.org/mesa/mesa/-/issues/10159

**etnaviv** driver requires pycparser to build

Package updates:
- spirv-tools: enable host build- spirv-tools: move package.mk from vulkan to graphics
- spirv-headers: move package.mk from vulkan to graphics
- spirv-llvm-translator: initial package
- llvm: include build of llvm-spirv
- libclc: initial package
- libdrm: enable host build
- pycparser: initial package
- mesa: update to 24.1.0
- glslang: update unpack

```
=== tested on ===
Generic.x86_64-devel-20240525062558-d5259d1
Linux nuc12 6.9.1 #1 SMP Sat May 25 05:18:45 UTC 2024 x86_64 GNU/Linux
Starting Kodi (21.0 (21.0.0) Git:21.0-Omega). Platform: Linux x86 64-bit
Using Release Kodi x64
Kodi compiled 2024-05-25 by GCC 14.1.0 for Linux x86 64-bit version 6.9.1 (395521)
Running on LibreELEC (heitbaum): devel-20240525062558-d5259d1 13.0, kernel: Linux x86 64-bit version 6.9.1
FFmpeg version/source: 6.0.1
Host CPU: 12th Gen Intel(R) Core(TM) i7-1260P, 16 cores available
[    0.000000] DMI: Intel(R) Client Systems NUC12WSKi7/NUC12WSBi7, BIOS WSADL357.0092.2024.0202.1711 02/02/2024
CApplication::CreateGUI - trying to init gbm windowing system
CApplication::CreateGUI - using the gbm windowing system
EGL_VENDOR = Mesa Project
GL_RENDERER = Mesa Intel(R) Graphics (ADL GT2)
GL_VERSION = OpenGL ES 3.2 Mesa 24.1.0
libva info: VA-API version 1.21.0
vainfo: VA-API version: 1.21 (libva 2.21.0)
vainfo: Driver version: Intel iHD driver for Intel(R) Gen Graphics - 24.2.3 (54156d6712)
```

successful build on all targets

### CODE REVIEW

File: packages/devel/libclc/package.mk [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+PKG_PATCH_DIRS+=" $(get_pkg_directory llvm)/patches"
+PKG_TOOLCHAIN="cmake"
+
+unpack() {
+  mkdir -p ${PKG_BUILD}
+  tar --strip-components=1 -xf ${SOURCES}/llvm/llvm-${PKG_VERSION}.tar.xz -C ${PKG_BUILD}
+}
+
+pre_configure() {
+  PKG_CMAKE_SCRIPT=${PKG_BUILD}/libclc/CMakeLists.txt
```
Comment (CvH) [inline packages/devel/libclc/package.mk:20]:
```suggestion
  PKG_CMAKE_SCRIPT="${PKG_BUILD}/libclc/CMakeLists.txt"
```

File: packages/devel/libclc/package.mk [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+pre_configure_host() {
+  #LIBCLC_TARGETS_TO_BUILD="spirv-mesa3d-\;spirv64-mesa3d-"
+  LIBCLC_TARGETS_TO_BUILD="spirv64-mesa3d-"
+
+  mkdir -p ${PKG_BUILD}/.${HOST_NAME}
```
Comment (CvH) [inline packages/devel/libclc/package.mk:27]:
```suggestion
  mkdir -p "${PKG_BUILD}/.${HOST_NAME}"
```

File: packages/devel/libclc/package.mk [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  mkdir -p ${PKG_BUILD}
+  tar --strip-components=1 -xf ${SOURCES}/llvm/llvm-${PKG_VERSION}.tar.xz -C ${PKG_BUILD}
+}
+
+pre_configure() {
+  PKG_CMAKE_SCRIPT=${PKG_BUILD}/libclc/CMakeLists.txt
+}
+
+pre_configure_host() {
+  #LIBCLC_TARGETS_TO_BUILD="spirv-mesa3d-\;spirv64-mesa3d-"
```
Comment (CvH) [inline packages/devel/libclc/package.mk:24]:
```suggestion
```

File: packages/graphics/mesa/package.mk [hunk 4]
```diff
@@ -99,3 +120,8 @@ if [ "${VULKAN_SUPPORT}" = "yes" ]; then
 else
   PKG_MESON_OPTS_TARGET+=" -Dvulkan-drivers="
 fi
+
+makeinstall_host() {
+  mkdir -p ${TOOLCHAIN}/bin
```
Comment (CvH) [inline packages/graphics/mesa/package.mk:125]:
```suggestion
  mkdir -p "${TOOLCHAIN}/bin"
```

File: packages/graphics/mesa/package.mk [hunk 5]
```diff
@@ -99,3 +120,8 @@ if [ "${VULKAN_SUPPORT}" = "yes" ]; then
...
+    cp -a src/intel/compiler/intel_clc ${TOOLCHAIN}/bin
```
Comment (CvH) [inline packages/graphics/mesa/package.mk:126]:
```suggestion
    cp -a "src/intel/compiler/intel_clc" "${TOOLCHAIN}/bin"
```

File: packages/graphics/spirv-llvm-translator/package.mk [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: GPL-2.0-only
+# Copyright (C) 2024-present Team LibreELEC (https://libreelec.tv)
+
+PKG_NAME="spirv-llvm-translator"
+PKG_VERSION="18.1.0"
+PKG_SHA256="78a770eff24d5ffe2798479845adec4b909cbf058ddc55830ea00fa7d2c1698a"
+PKG_LICENSE="Apache-2.0"
+PKG_SITE="https://github.com/KhronosGroup/SPIRV-LLVM-Translator"
+PKG_URL="https://github.com/KhronosGroup/SPIRV-LLVM-Translator/archive/v${PKG_VERSION}.tar.gz"
+PKG_DEPENDS_TARGET=""
```
Comment (CvH) [inline packages/graphics/spirv-llvm-translator/package.mk:10]:
```suggestion
```

File: packages/graphics/spirv-llvm-translator/package.mk [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: GPL-2.0-only
+# Copyright (C) 2024-present Team LibreELEC (https://libreelec.tv)
+
+PKG_NAME="spirv-llvm-translator"
+PKG_VERSION="18.1.0"
+PKG_SHA256="78a770eff24d5ffe2798479845adec4b909cbf058ddc55830ea00fa7d2c1698a"
+PKG_LICENSE="Apache-2.0"
```
Comment (CvH) [inline packages/graphics/spirv-llvm-translator/package.mk:7]: are you sure ? looks like another licence clusterfuck looking at the licence.txt at the git

File: packages/graphics/spirv-tools/package.mk [hunk 8]
```diff
@@ -11,6 +11,13 @@ PKG_SHA256="03ee1a2c06f3b61008478f4abe9423454e53e580b9488b47c8071547c6a9db47"
...
+PKG_DEPENDS_HOST="toolchain:host"
+PKG_DEPENDS_UNPACK="spirv-headers"
 PKG_LONGDESC="The SPIR-V Tools project provides an API and commands for processing SPIR-V modules."
-PKG_TOOLCHAIN="manual"
+
+post_unpack() {
+  mkdir -p ${PKG_BUILD}/external/spirv-headers
+    tar --strip-components=1 \
+      -xf ${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz \
+      -C ${PKG_BUILD}/external/spirv-headers
```
Comment (CvH) [inline packages/graphics/spirv-tools/package.mk:22]:
```suggestion
      -xf "${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz" \
      -C "${PKG_BUILD}/external/spirv-headers"
```

File: packages/graphics/vulkan/glslang/package.mk [hunk 9]
```diff
@@ -28,9 +28,14 @@ PKG_CMAKE_OPTS_COMMON="-DBUILD_EXTERNAL=ON \
 
 post_unpack() {
   # Enables SPIR-V optimzer capability needed for ENABLE_OPT CMake build option
+  mkdir -p ${PKG_BUILD}/External/spirv-tools
+    tar --strip-components=1 \
+      -xf ${SOURCES}/spirv-tools/spirv-tools-$(get_pkg_version spirv-tools).tar.gz \
+      -C ${PKG_BUILD}/External/spirv-tools
```
Comment (CvH) [inline packages/graphics/vulkan/glslang/package.mk:34]:
```suggestion
      -xf "${SOURCES}/spirv-tools/spirv-tools-$(get_pkg_version spirv-tools).tar.gz" \
      -C "${PKG_BUILD}/External/spirv-tools"
```

File: packages/graphics/vulkan/glslang/package.mk [hunk 10]
```diff
@@ -28,9 +28,14 @@ PKG_CMAKE_OPTS_COMMON="-DBUILD_EXTERNAL=ON \
...
   mkdir -p ${PKG_BUILD}/External/spirv-tools/external/spirv-headers
-    cp -R $(get_build_dir spirv-tools)/* ${PKG_BUILD}/External/spirv-tools
-    cp -R $(get_build_dir spirv-headers)/* ${PKG_BUILD}/External/spirv-tools/external/spirv-headers
+    tar --strip-components=1 \
+      -xf ${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz \
+      -C ${PKG_BUILD}/External/spirv-tools/external/spirv-headers
```
Comment (CvH) [inline packages/graphics/vulkan/glslang/package.mk:38]:
```suggestion
      -xf "${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz" \
      -C "${PKG_BUILD}/External/spirv-tools/external/spirv-headers"
```

File: packages/lang/llvm/package.mk [hunk 11]
```diff
@@ -41,8 +45,21 @@ PKG_CMAKE_OPTS_COMMON="-DLLVM_INCLUDE_TOOLS=ON \
...
 
+post_unpack() {
+  if listcontains "${GRAPHIC_DRIVERS}" "iris"; then
+    mkdir -p ${PKG_BUILD}/llvm/projects/{SPIRV-Headers,SPIRV-LLVM-Translator}
+      tar --strip-components=1 \
+        -xf ${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz \
+        -C ${PKG_BUILD}/llvm/projects/SPIRV-Headers
+      tar --strip-components=1 \
+        -xf ${SOURCES}/spirv-llvm-translator/spirv-llvm-translator-$(get_pkg_version spirv-llvm-translator).tar.gz \
+        -C ${PKG_BUILD}/llvm/projects/SPIRV-LLVM-Translator
```
Comment (CvH) [inline packages/lang/llvm/package.mk:59]:
```suggestion
    mkdir -p "${PKG_BUILD}/llvm/projects/{SPIRV-Headers,SPIRV-LLVM-Translator}"
      tar --strip-components=1 \
        -xf "${SOURCES}/spirv-headers/spirv-headers-$(get_pkg_version spirv-headers).tar.gz" \
        -C "${PKG_BUILD}/llvm/projects/SPIRV-Headers"
      tar --strip-components=1 \
        -xf "${SOURCES}/spirv-llvm-translator/spirv-llvm-translator-$(get_pkg_version spirv-llvm-translator).tar.gz" \
        -C "${PKG_BUILD}/llvm/projects/SPIRV-LLVM-Translator"
```

File: packages/lang/llvm/package.mk [hunk 12]
```diff
@@ -84,12 +101,20 @@ pre_configure_host() {
...
+  fi
 }
 
 post_makeinstall_host() {
   mkdir -p ${TOOLCHAIN}/bin
     cp -a bin/llvm-config ${TOOLCHAIN}/bin
     cp -a bin/llvm-tblgen ${TOOLCHAIN}/bin
+
+  if listcontains "${GRAPHIC_DRIVERS}" "iris"; then
+    cp -a bin/{llvm-as,llvm-link,llvm-spirv,opt} ${TOOLCHAIN}/bin
```
Comment (CvH) [inline packages/lang/llvm/package.mk:116]:
```suggestion
    cp -a "bin/{llvm-as,llvm-link,llvm-spirv,opt}" "${TOOLCHAIN}/bin"
```

File: packages/python/devel/pycparser/package.mk [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+PKG_LICENSE="BSD-3-Clause"
+PKG_SITE="https://pypi.org/project/pycparser/"
+PKG_URL="https://files.pythonhosted.org/packages/source/${PKG_NAME:0:1}/${PKG_NAME}/${PKG_NAME}-${PKG_VERSION}.tar.gz"
+PKG_DEPENDS_HOST="Python3:host setuptools:host"
+PKG_LONGDESC="Complete C99 parser in pure Python"
+PKG_TOOLCHAIN="manual"
+
+makeinstall_host() {
+  export DONT_BUILD_LEGACY_PYC=1
+  exec_thread_safe python3 setup.py install --prefix=${TOOLCHAIN}
```
Comment (CvH) [inline packages/python/devel/pycparser/package.mk:16]:
```suggestion
  exec_thread_safe python3 setup.py install --prefix="${TOOLCHAIN}"
```

**Review by CvH — CHANGES_REQUESTED** [review]
the changes are untested, feel free to disagree :)

File: packages/graphics/spirv-llvm-translator/package.mk (hunk 7 above)
Comment (heitbaum) [reply packages/graphics/spirv-llvm-translator/package.mk:7]: updated to included License file.

**Comment by heitbaum** [comment]
> the changes are untested, feel free to disagree :)

included most. some of the "" dont work e.g. mkdir "abc/{abc,def}" but otherwise all were ok.

**Review by CvH — APPROVED** [review]

**Pull request merged**